tests/query/joins/test_branch_cutter.py

```diff
@@ -0,0 +1,314 @@
+import pytest
+from snuba.query.expressions import (
+    Column,
+    CurriedFunctionCall,
+    Expression,
+    FunctionCall,
+    Literal,
+    SubscriptableReference,
+)
+from snuba.query.joins.classifier import (
+    BranchCutter,
+    MainQueryExpression,
+    SubExpression,
+    SubqueryExpression,
+    UnclassifiedExpression,
+)
+
+TEST_CASES = [
```

Comment:
Many of these scenarios do not require subqueries and hence branch cutting to work. Are you planning to run every single query through the BranchCutter, or only when we detect that more than one entity/table is being referenced?

Reply:
It will be every single root expression in the main query to decide what to send to the subqueries.
The BranchCutter deals with expressions, not with queries.
We will need to run it on everything because it is the BranchCutter that tells us what can be pushed down and what cannot.